### PR TITLE
TandemFoil: Fourier+physics+no-EMA long training T_max=300

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current TandemFoil best (82.65) used T_max=30 tuned for the 30-min budget. At that budget we only completed 14 epochs — and the model was still sharply improving (95.63 → 82.65 in the last 2 epochs). With the new 180-min budget and \`--epochs 999\`, we expect ~84 epochs of training.

The problem: T_max=30 at 84 epochs creates ~2.8 full cosine restart cycles. T_max=300 scales approximately to the new budget (roughly 1 full cycle per ~3.5 epochs), which may allow the optimizer to settle into a deeper minimum rather than oscillating through many fast restarts.

**Hypothesis:** T_max=300 is a better match to the 180-min training budget and may lead to better final convergence than T_max=30.

## Instructions

Run a single sequential training with T_max=300:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 3e-4 \
  --cosine_t_max 300 \
  --no-use-ema \
  --model_slices 64 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb_group fern-tandem-tmax300-longrun
```

**CRITICAL:** Always include `--epochs 999` — there is a bug where epochs=2 is the hardcoded default and SENPAI_MAX_EPOCHS only caps but does not raise it.

## Baseline

- **val_primary/surface_pressure_mae:** 82.65 (new best)
- **test_primary/surface_pressure_mae:** 80.63
- **Per-split val MAE:** single_in_dist=102.40, geom_camber_rc=88.97, geom_camber_cruise=62.37, re_rand=76.87
- **Best PR:** #2473 (edward — Fourier + physics + no-EMA, slices=64, T_max=30, Lion lr=3e-4, 14 epochs in 30 min)
- **W&B run:** nh380grv
- **Reproduce:** `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine_t_max 30 --no-use-ema --model_slices 64 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition`

Budget: 180 min

## Report

Please report:
- val_primary/surface_pressure_mae
- test_primary/surface_pressure_mae
- Epochs completed
- Per-split MAE breakdown (single_in_dist, geom_camber_rc, geom_camber_cruise, re_rand)
- W&B run ID

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### T_max=300 Long Training — **NEW BEST** ✓

| Metric | Baseline (T_max=30, #2473) | T_max=300 | Delta |
|--------|---------------------------|-----------|-------|
| **val_primary/surface_pressure_mae** | **82.65** | **80.50** | **-2.6%** |
| **test_primary/surface_pressure_mae** | **80.63** | **77.82** | **-3.5%** |
| val_single_in_dist | 102.40 | 91.68 | -10.5% |
| val_geom_camber_rc | 88.97 | 90.20 | +1.4% |
| val_geom_camber_cruise | 62.37 | 62.36 | ~0% |
| val_re_rand | 76.87 | 77.78 | +1.2% |

- **Epochs completed:** 14 (30-min timeout, same as baseline)
- **Best epoch:** 14 (at LR trough = 0)
- **W&B run ID:** 4ie6hkop
- **Peak VRAM:** ~87 GB (RTX PRO 6000 Blackwell)

### Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil --optimizer lion --lr 3e-4 \
  --cosine-t-max 300 --no-use-ema --model-slices 64 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --agent fern \
  --wandb-name "fern/tandem-tmax300-longrun" \
  --wandb-group fern-tandem-tmax300-longrun
```

### What happened

**T_max=300 beats the baseline** on val_primary (80.50 vs 82.65, -2.6%) and test_primary (77.82 vs 80.63, -3.5%). The improvement comes primarily from a large gain on `single_in_dist` (-10.5%), while other splits are roughly comparable.

**Key finding — LR cycling dynamics:** The cosine scheduler steps per-batch (not per-epoch). With 750 steps/epoch, T_max=300 creates ~1.25 full cosine cycles per epoch (period = 600 steps). This causes a characteristic oscillation pattern in val MAE:
- **LR troughs (lr=0):** Best val scores (epochs 2, 6, 10, 14)
- **LR peaks (lr=3e-4):** Temporary regressions (epochs 4, 8, 12)
- **Mid-cycle (lr=1.5e-4):** Intermediate recovery

Despite the oscillation, the trend converges — the model discovers deeper minima after each LR cycle. The trough progression: 154.39 → 97.10 → 86.22 → **80.50**.

**Comparison with T_max=30:** The baseline's T_max=30 creates ~12.5 cycles/epoch — the LR oscillates so fast that epochs effectively average over many cycles. T_max=300 gives the optimizer longer to settle per cycle, which appears to find better local minima despite the epoch-level volatility.

**Note on budget:** The PR hypothesized T_max=300 for a 180-min budget, but the actual env vars capped at 30 min / 50 epochs. The result is still meaningful — T_max=300's slower cycling helped at 14 epochs too.

### Suggested follow-ups

1. **T_max=150 or T_max=600:** Test intermediate and longer cycle lengths to find the optimal per-step T_max. The sweet spot may differ from 300.
2. **Per-epoch scheduler stepping:** The current per-step scheduling creates unintended intra-epoch cycling. Switching to per-epoch `scheduler.step()` would give T_max its intended meaning (cycles per epoch, not per step).
3. **CosineAnnealingWarmRestarts with larger T_0:** This could give the optimizer a full warm restart cycle spanning multiple epochs, potentially eliminating the epoch-level regression pattern.
4. **Extended 180-min training:** With the longer budget, T_max=300 per-step would give even more cycles to converge. The current 14-epoch run was still improving sharply.

**Comment by tcapelle** [comment]
ADVISOR: Merging — val=80.50 beats the old baseline of 82.65 by 2.6% (test: 77.82 vs 80.63, -3.5%). T_max=300 creates ~2.5 cycles/epoch; LR trough progression 154→97→86→80.50 shows consistent improvement. Still improving at cutoff — the oscillating trough pattern holds across T_max values. Good work, fern.
